src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -5219,61 +5215,134 @@ private static boolean needIncrement(MutableBigInteger mdivisor, int roundingMod
+    }
+
+    /**
+     * @param n a non-negative integer
+     * @return {@code 5^(2^n)}
+     */
+    private static BigInteger fiveToTwoToThe(int n) {
+        int i = Math.min(n, FIVE_TO_2_TO.length - 1);
+        BigInteger pow = FIVE_TO_2_TO[i];
+        for (; i < n; i++)
+            pow = pow.multiply(pow);
+
+        return pow;
+    }
+
+    private static final double LOG_5_OF_2 = 0.43067655807339306; // double closest to log5(2)
+
     /**
      * Remove insignificant trailing zeros from this
      * {@code BigInteger} value until the preferred scale is reached or no
      * more zeros can be removed.  If the preferred scale is less than
      * Integer.MIN_VALUE, all the trailing zeros will be removed.
+     * Assumes {@code intVal != 0}.
      *
      * @return new {@code BigDecimal} with a scale possibly reduced
      * to be closed to the preferred scale.
      * @throws ArithmeticException if scale overflows.
      */
     private static BigDecimal createAndStripZerosToMatchScale(BigInteger intVal, int scale, long preferredScale) {
+        // avoid overflow of scale - preferredScale
+        preferredScale = Math.clamp(preferredScale, Integer.MIN_VALUE - 1L, Integer.MAX_VALUE);
+        int powsOf2 = intVal.getLowestSetBit();
+        // scale - preferredScale >= remainingZeros >= max{n : (intVal % 10^n) == 0 && n <= scale - preferredScale}
+        // a multiple of 10^n must be a multiple of 2^n
+        long remainingZeros = Math.min(scale - preferredScale, powsOf2);
+        if (remainingZeros <= 0L)
+            return valueOf(intVal, scale, 0);
+
+        final int sign = intVal.signum;
+        if (sign < 0)
+            intVal = intVal.negate(); // speed up computation of shiftRight() and bitLength()
+
+        intVal = intVal.shiftRight(powsOf2); // remove powers of 2
+        // Let k = max{n : (intVal % 5^n) == 0}, m = max{n : 5^n <= intVal}, so m >= k.
+        // Let b = intVal.bitLength(). It can be shown that
+        // | b * LOG_5_OF_2 - b log5(2) | < 2^(-21) (fp viz. real arithmetic),
+        // which entails m <= maxPowsOf5 <= m + 1, where maxPowsOf5 is as below.
+        // Hence, maxPowsOf5 >= k.
+        long maxPowsOf5 = Math.round(intVal.bitLength() * LOG_5_OF_2);
+        remainingZeros = Math.min(remainingZeros, maxPowsOf5);
+
         BigInteger[] qr; // quotient-remainder pair
-        while (intVal.compareMagnitude(BigInteger.TEN) >= 0
-               && scale > preferredScale) {
-            if (intVal.testBit(0))
-                break; // odd number cannot end in 0
-            qr = intVal.divideAndRemainder(BigInteger.TEN);
-            if (qr[1].signum() != 0)
-                break; // non-0 remainder
-            intVal = qr[0];
-            scale = checkScale(intVal,(long) scale - 1); // could Overflow
+        // Remove 5^(2^i) from the factors of intVal, until 5^remainingZeros < 5^(2^i)
+        // (i.e., there are too few powers of five left to remove).
```

Comment:
It's a bit unclear what's meant here with "_too few_ powers of five left to remove".
Can we have something more explicit?

Reply:
It's hard for me to think of something more explicit than the mathematical definitions already present in the comments...

Reply:
Maybe something like "too few powers of five left to remove with respect to the maximum number of removable zeros in the original value of intVal", I can't think of anything better for now...

Reply:
OK, let me give it a try, maybe tomorrow.

Reply:
@rgiulietti Or maybe "i.e., 5^(2^i) is larger than the largest power of five that is still removable from intVal", could it be ok?

Reply:
While I intuitively understand, and I'm convinced of the clever algorithm, I'm struggling to find a proof, in particular to formulate a useful invariant for the first loop which seamlessly would bind with the second loop and its invariant.
I need to find more time for this.

Reply:
@rgiulietti Actually, an useful invariant for `remainingZeros` follows directly from its definition: letting `z = max{n : ((intVal * 2^powsOf2) % 10^n) == 0 && n <= scale - preferredScale}`, at each iteration, it is true that `remainingZeros >= z`, and the loop condition is `remainingZeros >= 2^i`. So, if `z < 2^i`, then `intVal % 5^(2^i) != 0` and the "then" branch of the "if" is executed, otherwise `z >= 2^i` and the "else" branch is executed. In both branchs, the invariant is preserved by the instructions.

Reply:
Yes, I considered that as well.
Not sure, however, if this is sufficient to prove that intVal has been indeed divided by 5^z after the 2nd loop is terminated.

Reply:
Anyway, if you find a nice proof please add it to the comments. The algorithm is quite sophisticated, so it needs one.

Reply:
I would say yes... The invariant `i == max{n : 5^(2^n) <= 5^remainingZeros}` should be a sufficient condition to affirm it, indeed the loop ends when `remainingZeros == 0`, so `remainingZeros >= z` implies `z == 0`...

Reply:
> Yes, I considered that as well. Not sure, however, if this is sufficient to prove that intVal has been indeed divided by 5^z after the 2nd loop is terminated.

New comments just added should be sufficient to prove it.

Reply:
The comments are OK. However, there seems to be no explicit relation between the "running" vars used in the reasoning and the expected outcome.

To clarify what I mean, let v0, z0 and s0 be the initial values at method entry of intVal, z and scale, resp.
At the exit of the method, the following is supposed to hold by contract:

v0 = intVal * 10^z0, s0 = scale + z0.

At least to me, this is not apparent from the comments.

Reply:
> The comments are OK. However, there seems to be no explicit relation between the "running" vars used in the reasoning and the expected outcome.
> 
> To clarify what I mean, let v0, z0 and s0 be the initial values at method entry of intVal, z and scale, resp. At the exit of the method, the following is supposed to hold by contract:
> 
> v0 = intVal * 10^z0, s0 = scale + z0.
> 
> At least to me, this is not apparent from the comments.

@rgiulietti Starting from the facts `min(scale - preferredScale, powsOf2) >= remainingZeros` and `remainingZeros >= 2^i`, if `intVal % 5^(2^i) == 0`, then `(intVal * 2^powsOf2) % 10^(2^i) == 0`, so dividing `intVal` by `5^(2^i)` is legit (in the sense that we don't remove more powers of 2 and we don't reduce the scale than the allowed) and exactly mimics the division `(intVal * 2^powsOf2) / 10^n` subtracting `2^i` from `powsOf2`. This allows to reduce the initial problem to the new values of the "running" vars after an iteration. So, I don't see the reason why it should not be apparent from the comments onestly...

Reply:
@rgiulietti That's correct, although I prefer that the correctness is proved by "reducing the size of the problem" rather than "total number of powers removed", because it was the logic to prove implicitly the correctness of the older implementation, namely by reducing the number of zeros without exceeding the largest scale decreasing allowed.

Reply:
With the definition remainingZeros = scale - preferredScale, the proof above could be adapted almost on the fly to the old implementation.

In any reduction approach, you still need to prove that the reduced problem is equivalent to the original. This is to say that the running variables (in code and proof) must somehow be related to the initial values.

Anyway, very nice contribution! Thanks.

Reply:
Will approve at the beginning of next week to let you add some last minute (not substantial) changes during the next few days.

Reply:
> In any reduction approach, you still need to prove that the reduced problem is equivalent to the original. This is to say that the running variables (in code and proof) must somehow be related to the initial values.

Yes, although in the proof by "reducing the size of the problem" this should be evident by induction...

Reply:
> Will approve at the beginning of next week to let you add some last minute (not substantial) changes during the next few days.

I thought that maybe could be useful add a comment that highlights that `(intVal * 2^powsOf2) / 10^z` and `scale - z` are invariant values, to point out the substructure of the problem better...